Sources/UIKit/PortalCollectionView.swift

```diff
@@ -0,0 +1,162 @@
+        let item = items[indexPath.row]
+        item.onTap |> { mailbox.dispatch(message: $0) }
+    }
+    
+    public func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
+        lastOffset = scrollView.contentOffset.x
+    }
+    
+    public func scrollViewWillEndDragging(_ scrollView: UIScrollView, withVelocity velocity: CGPoint,
+                                          targetContentOffset: UnsafeMutablePointer<CGPoint>) {
+        guard isSnapToCellEnabled else { return }
+        
+        let currentOffset = CGFloat(scrollView.contentOffset.x)
+        
+        if currentOffset == lastOffset {
+            return
+        }
+        
+        let lastPosition = selected
+        if currentOffset > lastOffset {
+            if lastPosition < items.count - 1 {
+                selected = lastPosition + 1 // Move to the right
+                scrollToItem(selected, animated: true)
+            }
+        } else if currentOffset < lastOffset {
+            if lastPosition >= 1 {
+                selected = lastPosition - 1
+                scrollToItem(selected, animated: true) // Move to the left
+            }
+        }
+    }
+    
+}
+
+fileprivate extension PortalCollectionView {
+    
+    fileprivate func scrollToItem(_ position: Int, animated: Bool) {
+        DispatchQueue.main.async(execute: {
+            let indexPath = IndexPath(item: position, section: 0)
+            self.scrollToItem(at: indexPath, at: .centeredHorizontally, animated: animated)
+        })
+    }
+
+    fileprivate func dequeueReusableCell(with identifier: String, for indexPath: IndexPath) -> PortalCollectionViewCell<MessageType> {
+        if let cell = dequeueReusableCell(withReuseIdentifier: identifier, for: indexPath) as? PortalCollectionViewCell<MessageType> {
+            cell.layoutEngine = layoutEngine
+            return cell
+        } else {
+            let cell = PortalCollectionViewCell<MessageType>(layoutEngine: layoutEngine)
+            cell.mailbox.forward(to: mailbox)
+            return cell
+        }
+    }
+    
+    fileprivate func itemRender(at indexPath: IndexPath) -> Component<MessageType> {
+        // TODO cache the result of calling renderer. Once the diff algorithm is implemented find a way to only
+        // replace items that have changed.
+        // IGListKit uses some library or algorithm to diff array. Maybe that can be used to make the array diff
+        // more efficient.
+        //
+        // https://github.com/Instagram/IGListKit
+        //
+        // Check the video of the talk that presents IGListKit to find the array diff algorithm.
+        // Also there is Dwifft which seems to be based in the same algorithm:
+        //
+        // https://github.com/jflinter/Dwifft
+        //
+        let item = items[indexPath.row]
+        return item.renderer()
+    }
+    
+}
+
+public struct CollectionViewLayoutValues {
```

Comment:
~~This strcut should be defined inside `Collection.swift` file and all dependencies with UIKit should be removed.~~ See my comment above about this type.